Add TranspileLayout to the C API

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds the TranspileLayout struct added in #14778 to the C
API. It is a straightforward addition to the API as the interface is
fairly simple. Given a pointer to a TranspileLayout the methods return
the arrays for the various views of the layout or the number of qubits
for the input and output circuits.

One small change was made to the PhysicalQubit and VirtualQubit type
definitions to use #[repr(transparent)] which in practice shouldn't
change anything, but it makes it explicit that these can be used
interchangeably with a u32. This simplifies how we can return arrays of
either type to C because it is safe to cast them as pointers to u32s.

Note that nothing in the C API is capable of generating a transpile
layout currently. The intent is for it to be part of the full
transpile() function return (see #14778). In the meantime this is
tested solely through rust tests and there aren't any tests written
in C yet. A potential follow up commit can be made to update the
standalone vf2 layout pass function to use a transpile layout inside
it's result type which would be more ergonomic than it's current
layout object in the result. But to keep this PR targeted this doesn't
update that interface and only exposes the struct to the C API.

### Details and comments

~This is based on #14778 and this PR will need to be rebased after #14778 merges. In the meantime you can view the content of this PR with the HEAD commit: https://github.com/Qiskit/qiskit/pull/14780/commits/8efe505f5c1adae7f880ba6910ebf046633102a9~ This has been rebased now

mtreinish committed:
Add TranspileLayout to the C API

This commit adds the TranspileLayout struct added in #14778 to the C
API. It is a straightforward addition to the API as the interface is
fairly simple. Given a pointer to a TranspileLayout the methods return
the arrays for the various views of the layout or the number of qubits
for the input and output circuits.

One small change was made to the PhysicalQubit and VirtualQubit type
definitions to use #[repr(transparent)] which in practice shouldn't
change anything, but it makes it explicit that these can be used
interchangeably with a u32. This simplifies how we can return arrays of
either type to C because it is safe to cast them as pointers to u32s.

Note that nothing in the C API is capable of generating a transpile
layout currently. The intent is for it to be part of the full
transpile() function return (see #14778). In the meantime this is
tested solely through rust tests and there aren't any tests written
in C yet. A potential follow up commit can be made to update the
standalone vf2 layout pass function to use a transpile layout inside
it's result type which would be more ergonomic than it's current
layout object in the result. But to keep this PR targeted this doesn't
update that interface and only exposes the struct to the C API.

mtreinish commented:
Now that https://github.com/Qiskit/qiskit/pull/14778 has merged this has been rebased and is unblocked

jakelishman commented on crates/cext/src/transpiler/transpile_layout.rs:
```diff
@@ -0,0 +1,337 @@
+    /// A pointer to an array of ``uint32_t`` representing the physical qubit
+    /// indices.
+    array: *mut u32,
+}
```
If we keep this, I think we should generalise it to `QkListU32` or something, since I imagine we're going to be dumping `Box<[T]>` back to C in more than one place.

That said, in all cases for `QkTranspileLayout`, the size that we'll write out is going to be trivially knowable by the caller.  Perhaps it's better for the functions to take a mutable caller-provided buffer and document the size it must be?

mtreinish replied:
With the filter ancillas option I felt like the array type return seemed like a better choice because the size will vary based on the argument values. While it's in the user's control it felt more error prone to do something like:
```c
u32 *layout_array = malloc(4 * qk_transpile_layout_num_input_qubits(layout));
qk_transpile_layout_initial_layout(layout, layout_array, true)
```
and 
```c
u32 *layout_array = malloc(4 * qk_transpile_layout_num_output_qubits(layout));
qk_transpile_layout_initial_layout(layout, layout_array, false)
```
where you have to match the filter_ancillas argument value with the right size.

jakelishman replied:
Letting the caller provide the buffer means that it doesn't have to be reallocated on every call, and it's one less thing that needs to be freed specially (since it could even be done on the stack in theory).

The caller can easily provide a buffer that's bigger than necessary - it wouldn't be a bug if you malloc'd the number of output qubits. I feel like we should tend towards C-sided memory allocation whenever reasonable - it means less cleanup and tighter control over performance.

Cryoris replied:
The intermediate solution would be to take a user-allocated `*mut QkLayoutArray` as input and write in that buffer. We could assert that `size` has the right value (ok it doesn't prove that the buffer has the right size, but it's a hint) and the user could re-use the memory. 

Also we have the same convention for "getting stuff" by mutable ref in the `QkCircuit` and `QkObs`.

jakelishman replied:
This isn't really the same as `QkCircuit` and `QkObs`, which both to allocate structured internal objects with a lot of internal detail about the number of elements.  This action is far simpler and we're making it more complex by introducing a new structure that has to be manually freed with non-standard tooling: "fill a buffer with `n` ints", where `n` is chosen to be either the number of input qubits or the number of output qubits with an argument.  That's a normal kind of C API - you typically bring your own buffer for that kind of that.  There's no need for Rust to allocate one for you.

jakelishman replied:
Also, `num_input_qubits` and `num_output_qubits` are typically going to be known by the user without having to make any call on the layout object, because they're going to be the number of qubits they put in their circuit, and the number of qubits of the backend they're compiling against.  What's more, if you're _using_ this object, you probably already have a bunch of other structures that already need to know the length of what it's going to be before you have to make it, including other dynamic allocations.

mtreinish replied:
I reworked the interface in: https://github.com/Qiskit/qiskit/pull/14780/commits/c12428039a91b0abd1d180d70646b8793f9722df

jakelishman commented on crates/cext/src/transpiler/transpile_layout.rs:
```diff
@@ -0,0 +1,337 @@
+///
+/// @param layout A pointer to the ``QkTranspileLayout``.
+/// @param filter_ancillas If set to true the output array will not include any indicies for any
+/// ancillas added by the transpiler.
```
This should explain how to actually interpret the result.

jakelishman commented on crates/cext/src/transpiler/transpile_layout.rs:
```diff
@@ -0,0 +1,337 @@
+///
+/// # Safety
+/// Behavior is undefined if ``layout`` is not a valid, non-null pointer to a
+/// ``QkTranspileLayout``.
```
This also needs to explain how to interpret the result.

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/14780/commits/c12428039a91b0abd1d180d70646b8793f9722df

jakelishman commented on docs/cdoc/index.h:
```diff
@@ -45,3 +45,7 @@
 /**
  * @defgroup QkElidePermutationsResult QkElidePermutationsResult
  */
+
+/**
+ * @defgroup QkTranspileLayout QkTranspileLayout
+ */
```
Note to self: I suspect I've not done the docs correctly in some of the VF2 C API stuff I added.

jakelishman commented on docs/cdoc/qk-transpiler-layout.rst:
```diff
@@ -0,0 +1,100 @@
+
+   typedef struct QkTranspileLayout QkTranspileLayout
+
+The ``QkTranspileLayout`` type is used to module the permutations introduced by
```
```suggestion
The ``QkTranspileLayout`` type is used to model the permutations introduced by
```

jakelishman commented on docs/cdoc/qk-transpiler-layout.rst:
```diff
@@ -0,0 +1,100 @@
+circuit qubits to physical qubits on the target) and the routing permtuations
+are caused by swap gate insertion or permutation ellision prior to the initial
+layout being set in the transpiler pipeline. This type tracks these details and
+provide an interface to reason about these permutations.
```
It'd probably be good to either centralise this non-programming high-level discussion, or at least link from here to the version of it in the Python-space docs (and improve it if necessary).

jakelishman commented on docs/cdoc/qk-transpiler-layout.rst:
```diff
@@ -0,0 +1,101 @@
+initial layout and output permutations. The initial layout permutation is
+caused by setting and applying the initial layout (the mapping from virtual
+circuit qubits to physical qubits on the target) and the output permtuations
+are caused by swap gate insertion or permutation ellision prior to the initial
```
```suggestion
circuit qubits to physical qubits on the target) and the output permutation
is caused by swap gate insertion or permutation elision prior to the initial
```

jakelishman commented on docs/cdoc/qk-transpiler-layout.rst:
```diff
@@ -0,0 +1,101 @@
+circuit qubits to physical qubits on the target) and the output permtuations
+are caused by swap gate insertion or permutation ellision prior to the initial
+layout being set in the transpiler pipeline. This type tracks these details and
+provide an interface to reason about these permutations.
```
It'd be good to either link to the version of this discussion in Python space, or centralise both.

mtreinish replied:
I'm fine doing that, which doc would you suggest I link to?

jakelishman replied:
tbh I'm not sure we have a really good version of it in Python space yet, so let's just leave it be.  I've had Intentions™ for a few months to write them, but never actually got it done.  I'll revisit this when I do.

jakelishman commented on docs/cdoc/qk-transpiler-layout.rst:
```diff
@@ -0,0 +1,101 @@
+circuit qubits to physical qubits on the target) and the output permtuations
+are caused by swap gate insertion or permutation ellision prior to the initial
+layout being set in the transpiler pipeline. This type tracks these details and
+provide an interface to reason about these permutations.
```
Please can we not refer to the layout as "a permutation" in new docs?  It's a relabelling - our insistence that we treat it as a permutation is what leads to the crazy "implicit ancilla" system.

mtreinish replied:
Sure I can change the wording here

mtreinish replied:
Reworded in: https://github.com/Qiskit/qiskit/pull/14780/commits/70251e491667649ef413472f968f1dbf32bedfd9

jakelishman commented on docs/cdoc/qk-transpiler-layout.rst:
```diff
@@ -0,0 +1,101 @@
+circuit to the physical circuits as:
+
+0 -> 2, 1 -> 1, 2 -> 0
+
```
This presumably doesn't format nicely - perhaps:
```suggestion

.. code-block:: text
    0 -> 2, 1 -> 1, 2 -> 0

```

jakelishman commented on docs/cdoc/qk-transpiler-layout.rst:
```diff
@@ -0,0 +1,101 @@
+qubit 0 in the input circuit is now physical qubit 2, virtual qubit 1 in the input circuit is
+physical qubit 0, and virtual qubit 2 in the input circuit is physical qubit 1.
+
+The transpiler will also allocate ancillas qubits to the circuit if the target
```
```suggestion
The transpiler will also allocate ancilla qubits to the circuit if the target
```

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Jake Lishman <jake@binhbar.com>

Cryoris commented on crates/cext/src/transpiler/transpile_layout.rs:
```diff
@@ -0,0 +1,337 @@
+/// @ingroup QkTranspileLayout
+/// Return the number of qubits in the input circuit to the transpiler.
+///
+/// @param layout A pointer to the ``QkTranspileLayout``.
```
These first 2 functions miss the `@return` doc 🙂

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/14780/commits/c12428039a91b0abd1d180d70646b8793f9722df

Cryoris commented on crates/cext/src/transpiler/transpile_layout.rs:
```diff
@@ -0,0 +1,337 @@
+/// ``qk_transpile_layout_initial_layout()``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_layout_array_free(layout_array: QkLayoutArray) {
```
In our current setup `..._free` always frees a pointer, while `_clear` clears out a struct (like for `QkCircuitInstruction`). Here we're doing the latter, right?

mtreinish replied:
It's passed the `QkLayoutArray` by value not by pointer so it is freeing the object too

mtreinish committed:
Update high level docs

mtreinish committed:
Rework C interface to write arrays to user provided pointers

This commit reworks the C API for accessing the layout details to
writing the output to a user provided pointer rather than allocating the
output array and giving them the pointer and size of it.

mtreinish committed:
Merge branch 'main' into add-c-api-transpile-layout

mtreinish committed:
Fix lint and docs

mtreinish committed:
Add missing free method

Cryoris commented on docs/cdoc/qk-transpile-layout.rst:
```diff
@@ -0,0 +1,102 @@
+    uint32_t cx_0_qargs[2] = {0, 1};
+    qk_circuit_gate(qc, QKGate_CX, cx_0_qargs, NULL);
+    uint32_t cx_1_qargs[2] = {0, 2};
+    qk_circuit_gate(qc, QKGate_CX, cx_1_qargs, NULL);
```
```suggestion
    qk_circuit_gate(qc, QkGate_H, h_qargs, NULL);
    uint32_t cx_0_qargs[2] = {0, 1};
    qk_circuit_gate(qc, QkGate_CX, cx_0_qargs, NULL);
    uint32_t cx_1_qargs[2] = {0, 2};
    qk_circuit_gate(qc, QkGate_CX, cx_1_qargs, NULL);
```

Cryoris commented on docs/cdoc/qk-transpile-layout.rst:
```diff
@@ -0,0 +1,102 @@
+    uint32_t cx_0_qargs[2] = {2, 1};
+    qk_circuit_gate(qc, QKGate_CX, cx_0_qargs, NULL);
+    uint32_t cx_1_qargs[2] = {2, 0};
+    qk_circuit_gate(qc, QKGate_CX, cx_1_qargs, NULL);
```
```suggestion
    qk_circuit_gate(qc, QkGate_H, h_qargs, NULL);
    uint32_t cx_0_qargs[2] = {2, 1};
    qk_circuit_gate(qc, QkGate_CX, cx_0_qargs, NULL);
    uint32_t cx_1_qargs[2] = {2, 0};
    qk_circuit_gate(qc, QkGate_CX, cx_1_qargs, NULL);
```

Cryoris commented on docs/cdoc/qk-transpile-layout.rst:
```diff
@@ -0,0 +1,102 @@
+    uint32_t cx_1_qargs[2] = {2, 0};
+    qk_circuit_gate(qc, QKGate_CX, cx_1_qargs, NULL);
+
+then the result of ``qk_transpile_layout_initial_layout()`` will be an array:
```
The current way of writing makes it look like a method. What do you think about being explicit here? Or alternatively, maybe just removing the parenthesis? (Same for the other occurrences below)
```suggestion
then the result of ``qk_transpile_layout_initial_layout(layout)``,
where `layout` is the corresponding `QkTranspileLayout *`, will be an array:
```

mtreinish replied:
I'm not sure I understand what you mean by it looks like a method. I can drop the paren if you want, I'm not a fan of putting a single argument like the suggestion because the function takes more than a single argument.

mtreinish replied:
I dropped the paren in: https://github.com/Qiskit/qiskit/pull/14780/commits/78c0430755d096f8755f86036975fae81be2a799

Cryoris commented on docs/cdoc/qk-transpile-layout.rst:
```diff
@@ -0,0 +1,102 @@
+    uint32_t swap_qargs[2] = {1, 0};
+    qk_circuit_gate(qc, QKGate_Swap, swap_qargs, NULL);
+    uint32_t cx_1_qargs[2] = {2, 1};
+    qk_circuit_gate(qc, QKGate_CX, cx_1_qargs, NULL);
```
```suggestion
    qk_circuit_gate(qc, QkGate_H, h_qargs, NULL);
    uint32_t cx_0_qargs[2] = {2, 1};
    qk_circuit_gate(qc, QkGate_CX, cx_0_qargs, NULL);
    uint32_t swap_qargs[2] = {1, 0};
    qk_circuit_gate(qc, QkGate_Swap, swap_qargs, NULL);
    uint32_t cx_1_qargs[2] = {2, 1};
    qk_circuit_gate(qc, QkGate_CX, cx_1_qargs, NULL);
```

Cryoris commented on docs/cdoc/qk-transpile-layout.rst:
```diff
@@ -0,0 +1,102 @@
+this results in the output state of qubit 0 moving to qubit 1 and qubit 1's to qubit 0.
+This results in ``qk_transpile_layout_output_permutation()`` returning the array:
+``[1, 0, 2]`` to indicate this is the final position of each qubit's state after the
+initial layout mapping. If no swaps or permutation ellisions were made during the
```
1 or 2 `L`?
```suggestion
initial layout mapping. If no swaps or permutation elisions were made during the
```

Cryoris commented on docs/cdoc/qk-transpile-layout.rst:
nice docs page 👍🏻

Cryoris commented on crates/cext/src/transpiler/transpile_layout.rs:
```diff
@@ -0,0 +1,507 @@
+}
+
+/// @ingroup QkTranspileLayout
+/// Return the initial layout of a ``QkTranspileLayout``.
```
Would "query" be more accurate since it doesn't "return" the layout but instead writes it into the variable passed by the user? (Same for the other functions below)
```suggestion
/// Query the initial layout of a ``QkTranspileLayout``.
```

Cryoris reviewed (commented):
Some small comments otherwise LGTM 👍🏻

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Update crates/cext/src/transpiler/transpile_layout.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Query everything

mtreinish committed:
Remove paren from function names

Cryoris commented on crates/cext/cbindgen.toml:
```diff
@@ -54,3 +54,4 @@ prefix_with_name = true
 "TargetEntry" = "QkTargetEntry"
 "VF2LayoutResult" = "QkVF2LayoutResult"
 "ElidePermutationsResult" = "QkElidePermutationsResult"
+"TranspileLayout" = "QkTranspilerLayout"
```
In the docs you're calling the object `QkTranspileLayout` but here it's transpileR layout -- I guess we wanted `QkTranspileLayout`? 😄

mtreinish replied:
```suggestion
"TranspileLayout" = "QkTranspileLayout"
```

mtreinish replied:
I've made this mistake a ton of places my brain just autocompletes the "r" from navigating the code base

mtreinish committed:
Update crates/cext/cbindgen.toml

jakelishman commented on crates/cext/src/transpiler/transpile_layout.rs:
```diff
@@ -0,0 +1,507 @@
+        let result = layout.final_index_layout(true);
+        let expected = vec![PhysicalQubit(3), PhysicalQubit(5), PhysicalQubit(2)];
+        assert_eq!(expected, result);
+        let mut result = vec![u32::MAX; layout.num_input_qubits() as usize];
+        unsafe { qk_transpile_layout_final_layout(&layout, true, result.as_mut_ptr()) };
+        let expected = expected.into_iter().map(|x| x.0).collect::<Vec<_>>();
+        assert_eq!(expected.as_slice(), result.as_slice());
```
Nitpicky: it might have been a nice test to allocate the `Vec` longer than needed and then assert that we only wrote out the correct number of qubits.  But in theory, Miri should catch that anyway.

jakelishman reviewed (approved):
I'm happy to merge this as-is and get things unblocked for further development.
